File: 2. Baekjoon/코딩테스트/5. 누적합/main.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String args[]) throws Exception {
        // freopen
        System.setIn(new FileInputStream("input.txt"));

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        int arr[] = new int[n+1];
        st = new StringTokenizer(br.readLine());
        for (int i=1; i<=n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        int psum[] = new int[n+1];
        for (int i=1; i<=n; i++) {
            psum[i] = psum[i-1] + arr[i];
        }

        StringBuilder sb = new StringBuilder();
        for (int i=0; i<m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            // System.out.println(psum[b] - psum[a-1]);
            sb.append(psum[b] - psum[a-1]).append("\n");
        }
        System.out.println(sb);
    }
}

File: 2. Baekjoon/코딩테스트/5. 누적합/main.swift
// import Foundation

// freopen("input.txt", "r", stdin)
// let input = readLine()!.split(separator: " ").map { Int($0)! }
// let (n, m) = (input[0], input[1])
// let arr = [0] + readLine()!.split(separator: " ").map { Int($0)! }


// for _ in 0..<m {
//     let range = readLine()!.split(separator: " ").map { Int($0)! }
//     let (a, b) = (range[0], range[1])
//     print(psum[b] - psum[a-1])
// }































// //
// //  main.swift
// //  Algorithm
// //
// //  Created by 김동현 on 2/2/26.
// //

// /*
//  1<=N<=100,000
//  1<=M<=100,000
//  1<=A<=B<=N
 
//  // 1 2 3 4 5 6 7 8
 
//  구간쿼리 키워드 : psum, 팬윅트리 생각나야함
//  */

// import Foundation

// freopen("input.txt", "r", stdin)
// let input = readLine()!.split(separator: " ").map { Int($0)! }
// let (n, m) = (input[0], input[1])
// let arr = [0] + readLine()!.split(separator: " ").map { Int($0)! }

// // 처음 접근 방법
// //for _ in 0..<m {
// //    let range = readLine()!.split(separator: " ").map { Int($0)! }
// //    let (a, b) = (range[0], range[1])
// //    var sum: Int = 0
// //    
// //    // 시간복잡도 최악: 10만 * 1만 = 100억...
// //    for i in a...b {
// //        sum += arr[i]
// //    }
// //    print(a, b, sum)
// //}




// 누적합 방법
// 1-based index 맞추기

import Foundation
freopen("input.txt", "r", stdin)
let input = readLine()!.split(separator: " ").map { Int($0)! }
let (n, m) = (input[0], input[1])
let arr = [0] + readLine()!.split(separator: " ").map { Int($0)! }
var psum = Array(repeating: 0, count: n+1)

for i in 1...n {
    psum[i] = psum[i-1] + arr[i]
}

for _ in 0..<m {
    let range = readLine()!.split(separator: " ").map { Int($0)! }
    let (a, b) = (range[0], range[1])
    print(psum[b] - psum[a-1])
}
